toxcore/forwarding.c

```diff
@@ -0,0 +1,377 @@
+                                       const uint8_t *packet, uint16_t length)
+{
+    if (length < FORWARD_REQUEST_MIN_PACKET_SIZE) {
+        return false;
+    }
+
+    const uint8_t *const public_key = packet + 1;
+    const uint8_t *const forward_data = packet + (1 + CRYPTO_PUBLIC_KEY_SIZE);
+    const uint16_t forward_data_len = length - (1 + CRYPTO_PUBLIC_KEY_SIZE);
+
+    if (TIMED_AUTH_SIZE + sendback_data_len > MAX_SENDBACK_SIZE ||
+            forward_data_len > MAX_FORWARD_DATA_SIZE) {
+        return false;
+    }
+
+    const uint16_t len = forwarding_packet_length(sendback_data_len, forward_data_len);
+    VLA(uint8_t, forwarding_packet, len);
+
+    create_forwarding_packet(forwarding, sendback_data, sendback_data_len, forward_data, forward_data_len,
+                             forwarding_packet);
+
+    return route_packet(forwarding->dht, public_key, forwarding_packet, len) == len;
+}
+
+non_null(1, 2) nullable(3, 5)
+static int handle_forward_request(void *object, const IP_Port *source, const uint8_t *packet, uint16_t length,
+                                  void *userdata)
+{
+    const Forwarding *forwarding = (const Forwarding *)object;
+
+    uint8_t sendback_data[1 + MAX_PACKED_IPPORT_SIZE];
+    sendback_data[0] = SENDBACK_IPPORT;
+
+    const int ipport_length = pack_ip_port(forwarding->log, sendback_data + 1, MAX_PACKED_IPPORT_SIZE, source);
+
+    if (ipport_length == -1) {
+        return 1;
+    }
+
+    return handle_forward_request_dht(forwarding, sendback_data, 1 + ipport_length, packet, length) ? 0 : 1;
+}
+
+#define MIN_NONEMPTY_SENDBACK_SIZE TIMED_AUTH_SIZE
+#define FORWARD_REPLY_MIN_PACKET_SIZE (1 + 1 + MIN_NONEMPTY_SENDBACK_SIZE)
+
+non_null(1, 2) nullable(3, 5)
+static int handle_forward_reply(void *object, const IP_Port *source, const uint8_t *packet, uint16_t length,
+                                void *userdata)
+{
+    const Forwarding *forwarding = (const Forwarding *)object;
+
+    if (length < FORWARD_REPLY_MIN_PACKET_SIZE) {
+        return 1;
+    }
+
+    const uint8_t sendback_len = packet[1];
+    const uint8_t *const sendback_auth = packet + 1 + 1;
+    const uint8_t *const sendback_data = sendback_auth + TIMED_AUTH_SIZE;
+
+    if (sendback_len > MAX_SENDBACK_SIZE) {
+        /* value 0xff is reserved for possible future expansion */
+        return 1;
+    }
+
+    if (sendback_len < TIMED_AUTH_SIZE + 1) {
+        return 1;
+    }
+
+    const uint16_t sendback_data_len = sendback_len - TIMED_AUTH_SIZE;
+
+    if (length < 1 + 1 + sendback_len) {
+        return 1;
+    }
+
+    const uint8_t *const to_forward = packet + (1 + 1 + sendback_len);
+    const uint16_t to_forward_len = length - (1 + 1 + sendback_len);
+
+    if (!check_timed_auth(forwarding->mono_time, SENDBACK_TIMEOUT, forwarding->hmac_key, sendback_data, sendback_data_len,
+                          sendback_auth)) {
+        return 1;
+    }
+
+    if (*sendback_data == SENDBACK_IPPORT) {
```

Comment:
Prefer `[0]` when indexing in array-like objects.

Reply:
done